urld0f681c558_farom57_AstroidDriver_tgz-pJ8-laazotea_indi_server_examples_MinimumINDIServerJ8/src/laazotea/indi/driver/INDIProperty.java

```diff
@@ -0,0 +1,615 @@
+   * @param message A message to be sent to the client.
+   * @return the opening XML Element &lt;setXXXVector&gt; for this Property.
+   */
+  protected abstract String getXMLPropertySetInit(String message);
+
+  /**
+   * Gets the closing XML Element &lt;/setXXXVector&gt; for this Property.
+   *
+   * @return the closing XML Element &lt;/setXXXVector&gt; for this Property.
+   */
+  protected abstract String getXMLPropertySetEnd();
+
+  /**
+   * Saves the property and its elements to a file. Ideal to later restore it on
+   * subsecuent executions of the driver.
+   *
+   * @throws IOException
+   */
+  private void saveToFile() throws IOException {
+    File propertiesDir = new File(PROPERTIES_DIR_NAME);
+
+    if (!propertiesDir.exists()) {
+      propertiesDir.mkdir();
+    }
+
+    if (propertiesDir.exists()) {
+      if (propertiesDir.isDirectory()) {
+        File file = new File(propertiesDir, getPropertyNameForFile());
+        FileOutputStream fos = new FileOutputStream(file);
+
+        ObjectOutputStream oos = new ObjectOutputStream(fos);
+
+        oos.writeObject(this);
+
+        oos.close();
+      }
+    }
+  }
+
+  /**
+   * Convenience method to get the name of the file for a particular property to
+   * be saved / retrieved.
+   *
+   * @param property The property
+   * @return A name for a file in which a property will be saved / restored
+   */
+  private String getPropertyNameForFile() {
+    return removeCharacters(getDriver().getName()) + "_" + getName() + ".prop";
+  }
+
+  /**
+   * Convenience method to get the name of the file for a particular property to
+   * be saved / retrieved.
+   *
+   * @param driver The driver of the property
+   * @param propertyName The name of the property
+   * @return A name for a file in which a property will be saved / restored
+   */
+  private static String getPropertyNameForFile(INDIDriver driver, String propertyName) {
+    return removeCharacters(driver.getName()) + "_" + propertyName + ".prop";
+  }
+
+  /**
+   * Removes all non letters / numbers from a String.
+   *
+   * @param str An input string
+   * @return The same <code>str</code>string without any non letter / numbers.
+   */
+  private static String removeCharacters(String str) {
+    return str.replaceAll("[^a-zA-Z0-9]", "");
+  }
+
+  /**
+   * Loads a property from a file.
+   *
+   * @param driver The driver which will include the property
+   * @param propertyName The name of the property to load
+   * @return A property loaded from a file
+   * @throws INDIException If there is some problem loading it (for example if
+   * the file does not exist)
+   */
+  protected static INDIProperty loadFromFile(INDIDriver driver, String propertyName) throws INDIException {
+    File propertiesDir = new File(PROPERTIES_DIR_NAME);
+
+    File file = new File(propertiesDir, getPropertyNameForFile(driver, propertyName));
+
+    INDIProperty prop = null;
+
+    try {
+      FileInputStream fis = new FileInputStream(file);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *fis*. The resource is referenced at lines 585, 587. The resource is closed at line 589. There are other execution paths that do not contain closure statements, for example, when ObjectInputStream.readObject throws an exception. Ensure *fis* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).